felix/fv/pre_dnat_test.go

```diff
@@ -229,6 +326,74 @@ var _ = infrastructure.DatastoreDescribe("pre-dnat with initialized Felix, 2 wor
+
+						err = waitForMetricToSettle(mangleRulesMetric)
+						Expect(err).NotTo(HaveOccurred(), "Timed out waiting for Prometheus (mangle) metric to settle")
+
+						err = waitForMetricToSettle(filterRulesMetric)
+						Expect(err).NotTo(HaveOccurred(), "Timed out waiting for Prometheus (filter) metric to settle")
+
+						baselineMangleTableIptablesSave := dumpMangleTable()
+						baselineMangleRulesMetric, err := mangleRulesMetric.Int()
+						Expect(err).NotTo(HaveOccurred(), "Failed to fetch baseline mangle rules metric")
+
+						baselineFilterTableIptablesSave := dumpFilterTable()
+						baselineFilterRulesMetric, err := filterRulesMetric.Int()
+						Expect(err).NotTo(HaveOccurred(), "Failed to fetch baseline filter rules metric")
+
+						// Add a new HEP.
+						hep := api.NewHostEndpoint()
+						hep.Name = "t0"
+						hep.Spec.Node = tc.Felixes[0].Hostname
+						hep.Labels = map[string]string{"abc123": "true"}
+						hep.Spec.InterfaceName = "*"
+						_, err = client.HostEndpoints().Create(utils.Ctx, hep, utils.NoOptions)
+						Expect(err).NotTo(HaveOccurred())
+
+						// Add GNP for the new HEP.
+						policy := api.NewGlobalNetworkPolicy()
+						policy.Name = "allow-ingress-8055-1"
+						order := float64(11)
+						policy.Spec.Order = &order
+						policy.Spec.PreDNAT = true
+						policy.Spec.ApplyOnForward = true
+						protocol := numorstring.ProtocolFromString("tcp")
+						testPort := numorstring.SinglePort(9999)
+						ports := numorstring.SinglePort(8055)
+						metricsPort := numorstring.SinglePort(9091)
+						policy.Spec.Ingress = []api.Rule{{
+							Action:   api.Allow,
+							Protocol: &protocol,
+							Destination: api.EntityRule{Ports: []numorstring.Port{
+								testPort,
+								ports,
+								metricsPort,
+							}},
+						}}
+						policy.Spec.Selector = "has(abc123)"
+						_, err = client.GlobalNetworkPolicies().Create(utils.Ctx, policy, utils.NoOptions)
+						Expect(err).NotTo(HaveOccurred(), "Couldn't create pre-DNAT GNP")
+
+						Expect(waitForIptablesToSettle()).NotTo(HaveOccurred(), "Timed out waiting for IPTables rules to settle")
+						curMangleRulesMetric, err := mangleRulesMetric.Int()
+						Expect(err).NotTo(HaveOccurred(), "Failed to fetch mangle rules prom metric")
+						curFilterRulesMetric, err := filterRulesMetric.Int()
+						Expect(err).NotTo(HaveOccurred(), "Failed to fetch filter rules prom metric")
+
+						mangleRulesMetricDelta := curMangleRulesMetric - baselineMangleRulesMetric
+						filterRulesMetricDelta := curFilterRulesMetric - baselineFilterRulesMetric
+
+						// The number of cali rules in the dumped tables should increase by the same delta as the metrics
+						Eventually(dumpMangleTable).Should(HaveLen(len(baselineMangleTableIptablesSave) + mangleRulesMetricDelta))
+						Eventually(dumpFilterTable).Should(HaveLen(len(baselineFilterTableIptablesSave) + filterRulesMetricDelta))
+					})
```

Comment:
I recommend doing some more steps after this:
* Change the type of the policy from pre-DNAT to normal.  That's the most fiddly change to handle since it goes from referenced in one table to referenced in the other.
* Delete the policy, should go back to the count with the hep.
* Delete the hep, should get back to the base count.